polaris/polaris/sep10auth/views.py

```diff
@@ -89,20 +90,114 @@ def _transaction_from_post_request(request):
     return _get_transaction_json(request.body)
 
 
-def _validate_envelope_xdr(envelope_xdr):
+def _validate_challenge_xdr(envelope_xdr):
     """
     Validate the provided TransactionEnvelope XDR (base64 string). Return the
     appropriate error if it fails, else the empty string.
+
+    verify_challenge_transaction() does not verify that the transaction's
+    signers are valid and have the total weight greater or equal to the
+    source account's medium threshold value, so this function does that as
+    well.
     """
     try:
         verify_challenge_transaction(
             challenge_transaction=envelope_xdr,
             server_account_id=settings.SIGNING_KEY,
             network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE,
         )
-    except InvalidSep10ChallengeError as err:
-        return str(err)
-    return ""
+    except InvalidSep10ChallengeError as e:
+        raise ValueError(str(e))
+
+    transaction_envelope = TransactionEnvelope.from_xdr(
+        envelope_xdr, settings.STELLAR_NETWORK_PASSPHRASE
+    )
+    # verify_challenge_transaction ensures transaction has at least
+    # one operation whose source is the client account.
+    source_account = transaction_envelope.transaction.operations[0].source
```

Comment:
💡 It'd be nice to take this line out of users of the SDK. Users need to know to grab the source account out of the manage data operation. In the Go SDK change we're adding a separate `read_challenge_transaction` function that does much of the same verification as the `verify_challenge_transaction` but doesn't verify client signatures. It can return the source account. It also means the server doesn't need to decode the transaction at all. It might be useful to do something similar in the Python SDK.
https://github.com/stellar/go/pull/2071/files#diff-f6a8a4abae844d51cc40cb628ba55d71R417-R427

Reply:
Ok, I'll add a comment to add a function to extract the source account from the envelope XDR in the SDK.